File: surf-cloud-core/src/main/java/dev/slne/surf/cloud/core/processors/AbstractLifecycleProcessor.java
package dev.slne.surf.cloud.core.processors;

import dev.slne.surf.cloud.api.lifecycle.SurfLifecycle;
import java.util.List;
import java.util.concurrent.ConcurrentHashMap;
import lombok.extern.flogger.Flogger;
import org.springframework.beans.BeansException;
import org.springframework.beans.factory.config.BeanPostProcessor;
import org.springframework.util.CollectionUtils;
import org.springframework.util.MultiValueMap;

@Flogger
public abstract class AbstractLifecycleProcessor implements BeanPostProcessor {

  private final MultiValueMap<Class<?>, SurfLifecycle> lifecycleMap = CollectionUtils.toMultiValueMap(
      new ConcurrentHashMap<>());

  @Override
  public Object postProcessAfterInitialization(Object bean, String beanName) throws BeansException {
    if (!(bean instanceof SurfLifecycle lifecycle)) {
      return bean;
    }

    final Class<?> providingClass = getProvidingClass(lifecycle);
    lifecycleMap.add(providingClass, lifecycle);

    return bean;
  }

  protected abstract Class<?> getProvidingClass(SurfLifecycle lifecycle);

  public final List<SurfLifecycle> getLifecycles(Class<?> providingClass) {
    return lifecycleMap.get(providingClass);
  }
}

File: surf-cloud-core/src/main/kotlin/dev/slne/surf/cloud/core/processors/AbstractLifecycleProcessor.kt
package dev.slne.surf.cloud.core.processors

import dev.slne.surf.cloud.api.lifecycle.SurfLifecycle
import dev.slne.surf.cloud.api.util.add
import dev.slne.surf.cloud.api.util.mutableObject2MultiObjectsMapOf
import org.springframework.beans.BeansException
import org.springframework.beans.factory.config.BeanPostProcessor
import kotlin.reflect.KClass

abstract class AbstractLifecycleProcessor : BeanPostProcessor {
    private val lifecycleMap = mutableObject2MultiObjectsMapOf<KClass<*>, SurfLifecycle>()

    @Throws(BeansException::class)
    override fun postProcessAfterInitialization(bean: Any, beanName: String): Any {
        if (bean !is SurfLifecycle) return bean

        val providingClass = getProvidingClass(bean)
        lifecycleMap.add(providingClass, bean)

        return bean
    }

    protected abstract fun getProvidingClass(lifecycle: SurfLifecycle): KClass<*>

    fun getLifecycles(providingClass: KClass<*>) = lifecycleMap[providingClass]
}
